charmhelpers/contrib/openstack/utils.py

```diff
@@ -2497,3 +2514,33 @@ def _inner_sequenced_functions(configs):
         return state, message
 
     return _inner_sequenced_functions
+
+
+SubordinatePackages = namedtuple('SubordinatePackages', ['install', 'purge'])
+
+
+def get_subordinate_release_packages(os_release, package_type='deb'):
+    """Iterate over subordinate relations and get package information.
+
+    :param os_release: OpenStack release to look for
+    :type os_release: str
+    :param package_type: Package type (one of 'deb' or 'snap')
+    :type package_type: str
+    :returns: Packages to install and packages to purge or None
+    :rtype: SubordinatePackages[set,set]
+    """
+    install = set()
+    purge = set()
+
+    for rdata in container_scoped_relation_get('releases-packages-map'):
+        rp_map = json.loads(rdata or '{}')
+        for release in sorted(rp_map.keys(), reverse=True):
+            if (CompareOpenStackReleases(release) <= os_release and
```

Comment:
I think this deserves a comment because the sorting with reverse=True and the comparison here aren't crystal-clear. Are we taking here the highest subordinate release smaller or equal the principal release? or are we taking the smallest subordinate release greater or equal the principal release? Also why? (I can't really decide what's best myself and why) thanks :)

EDIT: I have now read the unit tests, so I now understand that this will give the most recent earlier release info (i.e. what I called "the highest subordinate release smaller or equal the principal release") for each subordinate. But I'm not sure why this is good. Why not taking the exact same release info, and if not available, we would end up waiting for the subordinate to upgrade? I'm sure if I read through the linked bug and wrap my head around it I will understand but maybe this can be summed up in a few words in the code here? Thanks for your patience

Reply:
oh I think I know: say the subordinate is implemented today and gives info up to wallaby. then it works so well that we don't touch it for cycle X, we want the principle of release X to still regard the information provided for wallaby as valid. Am I right?

Reply:
In short, yes. Just like with the selection of templates or charm implementation, if nothing has changed since past release R we still want to use the information from past release R now. And since you're inclined to ask the question that is justification enough for me to add a comment, coming up.